fri/src/details.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::BTreeSet;
+
+use ark_ff::{FftField, PrimeField};
+use ark_poly::EvaluationDomain;
+use channel::FSChannel;
+use commitment_scheme::table_utils::RowCol;
+use sha3::Digest;
+
+use crate::stone_domain::get_field_element_at_index;
+use crate::{folder::MultiplicativeFriFolder, parameters::FriParameters};
+
+pub fn second_layer_queries_to_first_layer_queries(
+    query_indices: &[u64],
+    first_fri_step: usize,
+) -> Vec<u64> {
+    let first_layer_coset_size = 1 << first_fri_step;
+    let mut first_layer_queries = Vec::with_capacity(query_indices.len() * first_layer_coset_size);
+
+    for &idx in query_indices {
+        for i in (idx * first_layer_coset_size as u64)..((idx + 1) * first_layer_coset_size as u64)
+        {
+            first_layer_queries.push(i);
+        }
+    }
+
+    first_layer_queries
+}
+
+pub fn apply_fri_layers<F: FftField + PrimeField, E: EvaluationDomain<F>>(
```

Comment:
Might be cleaner to have a method applying a single round of FRI folding, then apply it many times. The current function might hide unintended O(n^2) behavior.